qutebrowser/completion/completionwidget.py

```diff
@@ -241,63 +238,75 @@ def completion_item_focus(self, which):
         selmodel.setCurrentIndex(
             idx, QItemSelectionModel.ClearAndSelect | QItemSelectionModel.Rows)
 
-    def set_model(self, model):
+        count = self.model().count()
+        if count == 0:
+            self.hide()
+        elif count == 1 and config.get('completion', 'quick-complete'):
+            self.hide()
+        elif config.get('completion', 'show') == 'auto':
+            self.show()
+
+    def set_model(self, model, pattern=None):
         """Switch completion to a new model.
 
         Called from on_update_completion().
 
         Args:
             model: The model to use.
+            pattern: The filter pattern to set (what the user entered).
         """
+        if model is None:
+            self._active = False
+            self.hide()
+            return
+
         old_model = self.model()
-        sel_model = self.selectionModel()
+        if model != old_model:
+            sel_model = self.selectionModel()
+
+            self.setModel(model)
+            self._active = True
 
-        self.setModel(model)
+            if sel_model is not None:
+                sel_model.deleteLater()
+            if old_model is not None:
+                old_model.deleteLater()
 
-        if sel_model is not None:
-            sel_model.deleteLater()
-        if old_model is not None:
-            old_model.deleteLater()
+        if (config.get('completion', 'show') == 'always' and
+                model.count() > 0):
```

Comment:
Right - I think pep8 used to require this and doesn't anymore, but I still find it much more readable.